moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.kt

```diff
@@ -0,0 +1,1169 @@
+      if (p == PEEKED_NONE) {
+        p = doPeek()
+      }
+      when (p) {
+        PEEKED_BEGIN_ARRAY -> {
+          pushScope(JsonScope.EMPTY_ARRAY)
+          count++
+        }
+        PEEKED_BEGIN_OBJECT -> {
+          pushScope(JsonScope.EMPTY_OBJECT)
+          count++
+        }
+        PEEKED_END_ARRAY -> {
+          count--
+          if (count < 0) {
+            throw JsonDataException("Expected a value but was ${peek()} at path $path")
+          }
+          stackSize--
+        }
+        PEEKED_END_OBJECT -> {
+          count--
+          if (count < 0) {
+            throw JsonDataException("Expected a value but was ${peek()} at path $path")
+          }
+          stackSize--
+        }
+        PEEKED_UNQUOTED_NAME, PEEKED_UNQUOTED -> skipUnquotedValue()
+        PEEKED_DOUBLE_QUOTED, PEEKED_DOUBLE_QUOTED_NAME -> skipQuotedValue(DOUBLE_QUOTE_OR_SLASH)
+        PEEKED_SINGLE_QUOTED, PEEKED_SINGLE_QUOTED_NAME -> skipQuotedValue(SINGLE_QUOTE_OR_SLASH)
+        PEEKED_NUMBER -> buffer.skip(peekedNumberLength.toLong())
+        PEEKED_EOF -> {
+          throw JsonDataException("Expected a value but was ${peek()} at path $path")
+        }
+      }
+      peeked = PEEKED_NONE
+    } while (count != 0)
+    pathIndices[stackSize - 1]++
+    pathNames[stackSize - 1] = "null"
+  }
+
+  @Throws(IOException::class)
+  override fun nextSource(): BufferedSource {
+    var p = peeked
+    if (p == PEEKED_NONE) {
+      p = doPeek()
+    }
+    var valueSourceStackSize = 0
+    val prefix = Buffer()
+    var state = JsonValueSource.STATE_END_OF_JSON
+    when (p) {
+      PEEKED_BEGIN_ARRAY -> {
+        prefix.writeUtf8("[")
+        state = JsonValueSource.STATE_JSON
+        valueSourceStackSize++
+      }
+      PEEKED_BEGIN_OBJECT -> {
+        prefix.writeUtf8("{")
+        state = JsonValueSource.STATE_JSON
+        valueSourceStackSize++
+      }
+      PEEKED_DOUBLE_QUOTED -> {
+        prefix.writeUtf8("\"")
+        state = JsonValueSource.STATE_DOUBLE_QUOTED
+      }
+      PEEKED_SINGLE_QUOTED -> {
+        prefix.writeUtf8("'")
+        state = JsonValueSource.STATE_SINGLE_QUOTED
+      }
+      PEEKED_NUMBER, PEEKED_LONG, PEEKED_UNQUOTED -> prefix.writeUtf8(nextString())
+      PEEKED_TRUE -> prefix.writeUtf8("true")
+      PEEKED_FALSE -> prefix.writeUtf8("false")
+      PEEKED_NULL -> prefix.writeUtf8("null")
+      PEEKED_BUFFERED -> {
+        val string = nextString()
+        JsonWriter.of(prefix).use { jsonWriter ->
+          jsonWriter.value(string)
+        }
+      }
+      else -> {
+        throw JsonDataException("Expected a value but was ${peek()} at path $path")
+      }
+    }
+
+    // Advance the path and clear peeked if we haven't already.
+    if (peeked != PEEKED_NONE) {
+      pathIndices[stackSize - 1]++
+      peeked = PEEKED_NONE
+    }
+    valueSource = JsonValueSource(source, prefix, state, valueSourceStackSize)
+    pushScope(JsonScope.STREAMING_VALUE)
+    return valueSource!!.buffer()
```

Comment:
We can eliminate this `!!` by holding the new instance in a local variable as well as writing it to `valueSource`.